### PR TITLE
[Fixes AB#194, fixes AB#403] Maps refactoring

### PR DESCRIPTION
**Pull request by Skafec**
Added maps with leaflet library to create project, update project and project details (Show location)

In project details checker is setup that checks if project location === undefined

### CODE REVIEW

**Review by mightymatth — CHANGES_REQUESTED** [review]
Merge the latest code from master, please. There are some of changes.

**Comment by Skafec** [comment]
> Merge the latest code from master, please. There are some of changes.

Sure, waiting for new functionality approval and then I will prepare branch for merge

**Comment by mj-ampnet** [comment]
@Skafec - you need to have the master merged into your branch with zero conflicts before a review is completed.

**Comment by Skafec** [comment]
New master is merged now

File: src/app/manage-projects/manage-single-project/manage-single-project.component.ts
Comment (Skafec) [reply src/app/manage-projects/manage-single-project/manage-single-project.component.ts:56]: We have to wait for backend response in order to get project location information.
Without it map breaks and isn't rendered (location is undefined)

Is there a better solution to init map once we get all the data we need?

File: src/app/manage-projects/manage-single-project/manage-single-project.component.ts
Comment (mightymatth) [inline src/app/manage-projects/manage-single-project/manage-single-project.component.ts:49]: Again, this is a hack. It would be better to catch the map event that signalizes that it's initialized, and then do something.

File: src/app/offers/offer-details/offer-details.component.html
Comment (mightymatth) [inline src/app/offers/offer-details/offer-details.component.html:106]: 😅

File: src/app/shared/services/location-map.service.ts
Comment (mightymatth) [inline src/app/shared/services/location-map.service.ts:6]: This should not be provided in `root`. You need a new instance of the service every time the component wants to use it. If you have 2 components on the same screen, they must not share this map object in the service. 
Instead, if you need to use the service, inject it in the component creation (`@Component(...)`). This way, every time the component is instantiated, it will get a new instance of the service.

```suggestion
@Injectable()
```

File: src/index.html
Comment (mightymatth) [inline src/index.html:13]: It looks like you added this line by yourself. This is not how Angular works. You add required CSS inside `angular.json`.

File: src/styles.css
Comment (mightymatth) [inline src/styles.css:2]: The same as in `src/index.html`. Please, check how other people do this, for example [here](https://www.geoapify.com/angular-leaflet-step-by-step-tutorial-to-add-a-map/).

File: src/app/shared/services/location-map.service.ts
Comment (mightymatth) [reply src/app/shared/services/location-map.service.ts:16]: Also, why are hardcoded coordinates here? Where is this location?

**Review by mightymatth — CHANGES_REQUESTED** [review]

File: src/app/offers/offer-details/offer-details.component.ts
Comment (mightymatth) [inline src/app/offers/offer-details/offer-details.component.ts:159]: Why these classes are not set inside `MapModalComponent`'s template?

File: src/app/projects/create-new-project/create-new-project.component.ts
Comment (mightymatth) [inline src/app/projects/create-new-project/create-new-project.component.ts:82]: Why these numbers? 🤓

**Review by mightymatth — CHANGES_REQUESTED** [review]

File: src/app/projects/create-new-project/create-new-project.component.ts
Comment (Skafec) [reply src/app/projects/create-new-project/create-new-project.component.ts:82]: These coordinates were used on old Google Maps as default location when user tries to create new project.

File: src/app/offers/offer-details/offer-details.component.ts
Comment (Skafec) [reply src/app/offers/offer-details/offer-details.component.ts:159]: I tried to add them on new HTML element, but modal would break.

Angular dynamically adds containers where these classes should go so
they have to be defined with bsModalRef

**Comment by Skafec** [comment]
Check please @mightymatth

**Review by mightymatth — COMMENTED** [review]
I made a bigger refactoring.
Look at [this](https://github.com/AMPnet/ampnet-crowdfunding-frontend/pull/74/commits/c8895bb350f2ebd200bbea525ee6431260f6a07b) commit, I'll leave comments there.

File: src/app/projects/create-new-project/create-new-project.component.html
Comment (mightymatth) [inline src/app/projects/create-new-project/create-new-project.component.html:33]: We tend to have a simple API for using our component. On your side, it's unclear what this component does with these events and `coordsSet`. 

What we need in this case is being able to attach latitude and longitude and being able to edit them.
With [Angular two-way binding](https://angular.io/guide/two-way-binding), we can directly map the changes. Check it out.

File: src/app/location-map/map-modal/map-modal.component.html
Comment (mightymatth) [inline src/app/location-map/map-modal/map-modal.component.html:9]: In this case, we don't need to edit the coordinates, so we don't use two-way binding. Notice the difference.

File: src/app/offers/offer-details/offer-details.component.ts
Comment (mightymatth) [inline src/app/offers/offer-details/offer-details.component.ts]: No need for two methods for this.

File: src/app/location-map/map-modal/map-modal.component.ts
Comment (mightymatth) [inline src/app/location-map/map-modal/map-modal.component.ts:21]: This seems to be a better approach to this issue. 
You made it from outside with `this.bsModalRef.content.rerenderMap.invalidateSize();`, but it's complicated to follow and why you do this to the next programmer that comes after you. If you do something for some edge case (e.g. 3rd-party bug), please make a comment on why you do it.

File: src/app/location-map/location-map.component.ts
Comment (mightymatth) [inline src/app/location-map/location-map.component.ts]:
```
        if (this.coordsSet.length !== 0) {
```
😁

File: src/app/location-map/location-map.component.ts
Comment (mightymatth) [inline src/app/location-map/location-map.component.ts]: Look at your previous code. You set a default marker location, put it on the map, then see if there are coordinates to set, and then set the marker another time. Not just unnecessary, but hard to read and could cause a map flickering.

**Comment by mightymatth** [comment]
Nothing else to change, I'll merge the latest master and set it to QA.

**Comment by mightymatth** [comment]
Ready for QA. @filipdujmusic

**Comment by filipdujmusic** [comment]
@mightymatth QA 🆗

**Review by mightymatth — APPROVED** [review]

**Pull request merged**
